### PR TITLE
[MXNET-978] [WIP] Higher Order Gradient Support `elemwise_mul`, `elemwise_div`.

### PR DESCRIPTION
**Pull request by kshitij12345**
## Description ##
PR intends to add support for higher order gradient for `elemwise_mul`, `elemwise_div`.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA-978 issue](https://issues.apache.org/jira/browse/MXNET-978) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] higher order gradient for a `elemwise_mul`, `elemwise_div`.
- [x] unit test for the same.

### CODE REVIEW

File: src/operator/tensor/elemwise_binary_op_basic.cc [hunk 1]
```diff
@@ -274,7 +298,43 @@ NNVM_REGISTER_OP(_backward_div)
...
+    auto r_y = MakeNode("reciprocal", n->attrs.name + "_r_y", {y}, nullptr, &n);
+    auto neg_two_r_y = MakeNode("_mul_scalar", n->attrs.name + "_neg_two_r_y",
+                            {nnvm::NodeEntry{r_y}}, &two, &n);
+    auto d2y = MakeNode("elemwise_mul", n->attrs.name + "_d2y",
+                            {nnvm::NodeEntry{neg_two_r_y}, dy}, &two, &n);
+
+    std::vector<nnvm::NodeEntry> ret;
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
+                              {ograds[0], nnvm::NodeEntry{dz_grad}}, nullptr, &n));
+    ret.emplace_back(MakeNode("zeros_like", n->attrs.name + "_backward_grad_grad_x",
```
Comment (apeforest) [inline src/operator/tensor/elemwise_binary_op_basic.cc]: Really?  Have you tested y = x * x?

File: src/operator/tensor/elemwise_binary_op_basic.cc [hunk 2]
```diff
@@ -251,7 +251,31 @@ NNVM_REGISTER_OP(_backward_mul)
...
+    // f(x, y) = x * y
+    // dx = z_grad * y, dy = z_grad * x
+    // d2x = 0, d2y = 0, dz_grad = dx + dy = y + x
+    auto dz_grad = MakeNode("elemwise_add", n->attrs.name + "_y_add_x",
+                            {n->inputs[2], n->inputs[1]}, nullptr, &n);
+
+    std::vector<nnvm::NodeEntry> ret;
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
+                              {ograds[0], nnvm::NodeEntry{dz_grad}}, nullptr, &n));
+    ret.emplace_back(MakeNode("zeros_like", n->attrs.name + "_backward_grad_grad_x",
```
Comment (apeforest) [inline src/operator/tensor/elemwise_binary_op_basic.cc]: Really? Have you tested z = x * x?

File: src/operator/tensor/elemwise_binary_op_basic.cc (hunk 1 above)
Comment (apeforest) [inline src/operator/tensor/elemwise_binary_op_basic.cc]: Same here. Please test z = x / (x^2 + 1)

File: src/operator/tensor/elemwise_binary_op_basic.cc (hunk 2 above)
Comment (kshitij12345) [reply src/operator/tensor/elemwise_binary_op_basic.cc]: Ah. Right. Thank You.

**Comment by roywei** [comment]
@mxnet-label-bot add [Operator, Backend]

**Comment by abhinavs95** [comment]
Adding WIP tag since its there in the title
@mxnet-label-bot add [Operator, Backend, pr-work-in-progress]

**Pull request closed without merging**